book/numerical_methods/integration1d.nim

```diff
@@ -0,0 +1,317 @@
+- `trapz`: Works for any data.
+-  `simpson`: Works for any data with 3 or more data points.
+- `romberg`: Works **only** for equally spaced points. The number of points must also be 
+of the form `2^n + 1` (eg. 3, 5, 9, 17, 33 etc).
+
+Luckily for us our data satisfies all of them ;) So let's get coding:
+  """
+
+  nbCode:
+    let exact = F(1) - F(0)
+
+    var trapzIntegral = trapz(ySample, xSample)
+    var simpsonIntegral = simpson(ySample, xSample)
+    var rombergIntegral = romberg(ySample, xSample)
+
+    echo "Exact:   ", exact
+    echo "Trapz:   ", trapzIntegral
+    echo "Simpson: ", simpsonIntegral
+    echo "Romberg: ", rombergIntegral
+    echo "Trapz Error:   ", abs(trapzIntegral - exact)
+    echo "Simpson Error: ", abs(simpsonIntegral - exact)
+    echo "Romberg Error: ", abs(rombergIntegral - exact)
+
+  nbText: md"""
+As expected all the methods underestimated the integral, but it might be unexpected that
+trapz performed the best out of them. Let's add a few more points, why not 33, and let's see if 
+that changes things!
+  """
+
+  nbCode:
+    xSample = numericalnim.linspace(0.0, 1.0, 33)
+    ySample = xSample.mapIt(f(it, nil))
+
+    sampledSpline = newHermiteSpline(xSample, ySample)
+    ySpline = sampledSpline.eval(xDense)
+
+    dfSample = seqsToDf(xSample, ySample, xDense, yDense, ySpline)
+    ggplot(dfSample) +
+      geom_point(aes("xSample", "ySample", color="Sampled")) +
+      geom_line(aes("xDense", "ySpline", color="Sampled")) +
+      geom_line(aes("xDense", "yDense", color="Dense")) +
+      scale_x_continuous() + scale_y_continuous() +
+      ggsave("images/sampledHumps33.png")
+
+    trapzIntegral = trapz(ySample, xSample)
+    simpsonIntegral = simpson(ySample, xSample)
+    rombergIntegral = romberg(ySample, xSample)
+
+    echo "Exact:   ", exact
+    echo "Trapz:   ", trapzIntegral
+    echo "Simpson: ", simpsonIntegral
+    echo "Romberg: ", rombergIntegral
+    echo "Trapz Error:   ", abs(trapzIntegral - exact)
+    echo "Simpson Error: ", abs(simpsonIntegral - exact)
+    echo "Romberg Error: ", abs(rombergIntegral - exact)
+
+  nbImage("images/sampledHumps33.png")
+
+  nbText: md"""
+As expected all methods became more accurate when we increased the amount of points.
+And from the graph we can see that the points capture the shape of the curve much better now.
+We can also note that simpson has overtaken trapz and romberg is neck-in-neck with trapz now.
+Experiment for yourself with different number of points, but asymtotically romberg will eventually
+beat simpson when enough points are used. 
+
+The take-away from this very limited testing is that depending on the characteristics and quality
+of the data, different methods might give the most accurate answer. Which one is hard to tell in general
+but trapz *might* be more robust for very sparse data as it doesn't "guess" as much as the others. But once again,
+it entirely depends on the data, so make sure to understand your data!  
```

Comment:
> so make sure to understand your data!  

:sob: reality doesn't want me to understand data sometimes though :rofl:

Reply:
😂 don't expect a computer to understand it better then 🤪